wallet: Faster transaction creation by removing pre-set-inputs fetching responsibility from Coin Selection

furszy opened the pull request:
#### # Context (Current Flow on Master)

In the transaction creation process, in order to select which coins the new transaction will spend,
we first obtain all the available coins known by the wallet, which means walking-through the
wallet txes map, gathering the ones that fulfill certain spendability requirements in a vector.

This coins vector is then provided to the Coin Selection process, which first checks if the user
has manually selected any input (which could be internal, aka known by the wallet, or external),
and if it does, it fetches them by searching each of them inside the wallet and/or inside the
Coin Control external tx data.

Then, after finding the pre-selected-inputs and gathering them in a vector, the Coin Selection
process walks-through the entire available coins vector once more just to erase coins that are
in both vectors. So the Coin Selection process doesn’t pick them twice (duplicate inputs inside
the same transaction).

#### # Process Workflow Changes

Now, a new method, `FetchCoins` will be responsible for:
1) Lookup the user pre-selected-inputs (which can be internal or external).
2) And, fetch the available coins in the wallet (excluding the already fetched ones).

Which will occur prior to the Coin Selection process. Which allows us to never include the
pre-selected-inputs inside the available coins vector in the first place, as well as doing other
nice improvements (written below).

So, Coin Selection can perform its main responsibility without mixing it with having to fetch
internal/external coins nor any slow and unneeded duplicate coins verification.

#### # Summarizing the Improvements:

1) If any pre-selected-input lookup fail, the process will return the error right away.
    (before, the wallet was fetching all the wallet available coins, walking through the
    entire txes map, and then failing for an invalid pre-selected-input inside SelectCoins)

2) The pre-selected-inputs lookup failure causes are properly described on the return error.
    (before, we were returning an "Insufficient Funds" error for everything, even if the failure
    was due a not solvable external input)

3) **Faster Coin Selection**: no longer need to "remove the pre-set inputs from the available coins
    vector so that Coin Selection doesn't pick them" (which meant to loop-over the entire
    available coins vector at Coin Selection time, erasing duplicate coins that were pre-selected).
   
    Now, the available coins vector, which is built after the pre-selected-inputs fetching,
    doesn’t include the already selected inputs in the first place.

4) **Faster transaction creation** for transactions that only use manually selected inputs.

    We now will return early, as soon as we finish fetching the pre-selected-inputs and
    not perform the resources expensive calculation of walking-through the entire wallet
    txes map to obtain the available coins (coins that we will not use).

---------------------------

Added a new bench (f6d0bb2) measuring the transaction creation process, for a wallet with ~250k UTXO, only using the pre-selected-inputs inside coin control. Setting `m_allow_other_inputs=false` to disallow the wallet to include coins automatically.

#### Result on this PR (tip f6d0bb2d):

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|        1,048,675.00 |              953.58 |    0.3% |      0.06 | `WalletCreateTransaction`

vs

#### Result on master (tip 4a4289e2):

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       96,373,458.20 |               10.38 |    0.2% |      5.30 | `WalletCreateTransaction`

The benchmark took to run in master: **96.37 milliseconds**, while in this PR: **1 millisecond**  🚀 .

furszy committed:
wallet: encapsulate pre-selected-inputs lookup into its own function

First step towards decoupling the pre-selected-inputs fetching functionality
from `SelectCoins`. Which, will let us not waste resources calculating the
available coins if one of the pre-set inputs has an error.

(right now, if one of the pre-set inputs is invalid, we first walk through
the entire wallet txes map just to end up failing right after it finish)

furszy commented:
rebased, conflicts solved.

ishaanam commented:
Concept ACK, this seems like a good improvement.

achow101 commented:
It's not clear to me that this provides any significant benefit over a much simpler change of just excluding pre-selected inputs from AvailableCoins. The only other improvement this provides is better error reporting, but I think it would be better to just go through SelectCoins and improve its error reporting instead of pulling things out so that errors are reported elsewhere.

furszy commented:
> It's not clear to me that this provides any significant benefit over a much simpler change of just excluding pre-selected inputs from AvailableCoins. The only other improvement this provides is better error reporting, but I think it would be better to just go through SelectCoins and improve its error reporting instead of pulling things out so that errors are reported elsewhere.

There are several benefits, (aside from what I wrote in the PR description):

#### Sources Architecture:

 - Do make sense for you to be fetching coins in a function whose main responsibility is perform the Coin Selection process?
 
    The Coin Selection should receive coins, and certain user/wallet parameters, so it can produce all the possible solutions for a certain target amount, picking and returning the best of them.

	Just imagine how this would looks like if Coin Selection would be a separated library (which, for its increasing specialization, could easily be one in the future). You would never include the pre-selected-inputs fetching and verification process into the Coin Selection responsibilities at all because you cannot/shouldn't do it there.

#### Tangible Improvements:

1.  As we will do the pre-selected-inputs fetching and verification prior to calling `AvailableCoins`, if there is an error when the wallet fetches any of the preset-inputs, the transaction creation process can fail right away without calling `AvailableCoins` (Function that, as we know, walks-through the entire wallet's transactions map and performs all the internal available coins checks).
 
     This prevents us from wasting resources that we are currently wasting for, literally, no reason.

	In master, we are first looping-through the entire wallet's map inside  `AvailableCoins`, which is a resource intensive task if you have a big wallet, just to fail right after it finishes, discarding the coins result, because one of the pre-set inputs has an error.

3.  If the user disallowed the "other inputs selection" by setting `m_allow_other_inputs=false` (which means: “no coins from the AvailableCoins result function will be used during Coin Selection”): 

	In master: we loop-through the entire wallet's map inside `AvailableCoins` first, then we discard the computed result (because we will not use it..), then we fetch and use the pre-selected-inputs only.
	Which is time and processing power wasted for no reason.

	In this PR: we simply select the preset inputs, skip `AvailableCoins` if `m_allow_other_inputs=false` and provides them to `SelectCoins`.

-------------------

> but I think it would be better to just go through SelectCoins and improve its error reporting instead of pulling things out so that errors are reported elsewhere.

We are currently failing inside the Coin Selection process due:
* “Invalid pre-selected input”
* “Not found pre-selected input”
* “Not solvable pre-selected input”

These errors have nothing to do with the Coin Selection process. Why keep them there?

It's not that I located these errors on a random place. I placed them inside a "Fetch Coins" procedure whose main responsibility is fetch the coins that the user pre-selected and fetch the available coins in the wallet. Which seems totally reasonable for me.

I mean, this errors have been obscured inside Coin Selection all this time, not being properly notified, because they should had never been placed there in the first place. It's all about a proper responsibility division to make the process clearer and cleaner so it's less error-prone (and in this case, in some occasions, actually faster and less resource consuming).

furszy committed:
bench: benchmark transaction creation process

Goal 1:
Benchmark the transaction creation process for pre-selected-inputs only.
Setting `m_allow_other_inputs=false` to disallow the wallet to include coins automatically.

Goal 2:
Benchmark the transaction creation process for pre-selected-inputs and coin selection.

-----------------------

Benchmark Setup:
1) Generates a 5k blockchain, loading the wallet with 5k transactions with two outputs each.
2) Fetch 4 random UTXO from the wallet's available coins and pre-select them as inputs inside CoinControl.

Benchmark (Goal 1):
Call `CreateTransaction` providing the coin control, who has set `m_allow_other_inputs=false` and
the manually selected coins.

Benchmark (Goal 2):
Call `CreateTransaction` providing the coin control, who has set `m_allow_other_inputs=true` and
the manually selected coins.

furszy commented:
To make the speed improvements more visible, pushed a new benchmark 0eb720f.

Basically, the new bench measures the transaction creation process, for a wallet with ~250k UTXO, only using the pre-selected-inputs inside coin control. Setting `m_allow_other_inputs=false` to disallow the wallet to include coins automatically.

#### This is the result on this PR (tip 0eb720f):

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|        1,048,675.00 |              953.58 |    0.3% |      0.06 | `WalletCreateTransaction`

vs

#### This is the result on current master (tip 4a4289e2):

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       96,373,458.20 |               10.38 |    0.2% |      5.30 | `WalletCreateTransaction`

#### Summary

The benchmark took to run in master: **96.37 milliseconds**, while in this PR: **1 millisecond**  🚀 .

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -105,6 +105,54 @@ void CoinsResult::clear()
 
+// Fetch the coin control selected inputs and wrap them inside a COutput.
+// Coins could be internal (from the wallet) or external.
+BResult<CoinsResult> FetchSelectedInputs(const CWallet& wallet, const CCoinControl& coin_control,
```
In 23448cae6f4a5f00766df9220860b16311266417 "wallet: encapsulate pre-selected-inputs lookup into its own function"

I think this should return a `SelectionResult` rather than `CoinsResult`. It would elimiate the second iteration inside of `SelectCoins`.

It could also return the selected value too so the caller could reduce the target before `SelectCoins` and eliminate an additional iteration of the preselcted inputs.

furszy replied:
good thought 👍🏼 

Most likely, will need to do this change incrementally (right after 23448cae) because if `FetchSelectedInputs` returns a `SelectionResult`, will need to add a method for merging two `SelectionResult` objects.

We are currently creating an `OutputGroup` at the beginning of `SelectCoins` with the pre-set-inputs just to merge it at the end of the function with the wallet's selection algorithm result. The reason is that `SelectResult` only has a function `AddInputs(OutputGroups)` and not an `AddInputs(set<COutputs>)` nor a merge method.

Side note, based on a quick code review: the selection result built from the combination of pre-set-inputs and the wallet selected ones should be having an inaccurate waste calculation -- will do some tests in another PR --).

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -153,6 +153,31 @@ BResult<CoinsResult> FetchSelectedInputs(const CWallet& wallet, const CCoinContr
     return result;
 }
 
+BResult<CoinsFund> FetchCoins(const CWallet& wallet, const CCoinControl& coin_control,
```
In 84b4ffe85ba0267c49080e9a096e45d6af63e720 "wallet: remove fetch pre-selected-inputs responsibility from SelectCoins"

I don't think it's necessary to have another function here to wrap these two calls. They could just be in `CreateTransactionInternal`

furszy replied:
True. It's not used anywhere else for now.
Have removed the function and the `CoinsFunds` struct as well.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -153,6 +153,31 @@ BResult<CoinsResult> FetchSelectedInputs(const CWallet& wallet, const CCoinContr
+        if (!result.pre_selected_inputs) return result.pre_selected_inputs.GetError();
+
+        // If no other inputs are allowed, return early
+        if (!coin_control.m_allow_other_inputs) return result;
```
M
In 84b4ffe85ba0267c49080e9a096e45d6af63e720 "wallet: remove fetch pre-selected-inputs responsibility from SelectCoins"

This check could be moved to the top of `AvailableCoins` and it would achieve the same effect. Doing so would also not change `AvailableCoins`'s semantics.

furszy replied:
I added it here (inside the `coin_control.HasSelected()` block) to not change the current behavior and minimize the amount of changes.

In master, even if `m_allow_other_inputs` is false, we are automatically falling back to select coins from the wallet if there are no pre-set inputs inside coin control.

This basically is because the default value for `m_allow_other_inputs` is false, and some RPC commands that don't allow inputs manual selection such as `sendtoaddress` and `sendmany` are not setting it to true. So.. they are always falling back to select coins from the wallet.

So.. if I move this check below, will need to set `m_allow_other_inputs=true` in every caller that does not add any input manually (or.. change the default value to be true inside coin control).

but.. let me see if those two RPC commands are the only affected ones. Maybe it's a straightforward change.
I do agree that the current master behavior of automatically falling back to select coins from the wallet even when `m_allow_other_inputs=false` is messy and not an expected behavior.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -575,24 +602,17 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
 {
     CAmount value_to_select = nTargetValue;
 
+    // Add the manually selected coins first
```
In 84b4ffe85ba0267c49080e9a096e45d6af63e720 "wallet: remove fetch pre-selected-inputs responsibility from SelectCoins"

I'm wondering if we should just make this function only return a `SelectionResult` for the automatic coin selection and move the preset input stuff to the caller. If `FetchSelectedInputs` made a `SelectionResult` for the preset inputs, then the only preset input things left in `SelectCoins` would be the merging of the `SelectionResult`s, and that can be done in the caller too.

furszy replied:
Been thinking about this and pushed some refactoring that you might like a bit more.

The idea was based on try to not mix the Coin Selection code with the inputs fetching while still be able to remove the manual coins selection stuff from the automatic coin selection, the extra loops for the pre-set inputs and  the `FetchCoins` wrapper function.

So, have decoupled the `SelectCoins` into two functions:

1) AutomaticCoinSelection:
Which is basically the previously called `SelectCoins` without the manually selected inputs stuff.
So, it will purely receive a set of coins and be in charge of selecting the best subset of them to
cover the target amount.

2) SelectCoins:
In charge of selecting all the user manually selected coins first ("pre-set inputs"), and
if coin_control 'm_allow_other_inputs=true', call 'AutomaticCoinSelection' to select a
set of coins such that the target - pre_set_inputs.total_amount is met. Then merge
both results.

furszy committed:
wallet: skip available coins fetch if "other inputs" are disallowed

no need to waste resources calculating the wallet available coins if
they are not going to be used.

The 'm_allow_other_inputs=true` default value change is to correct
an ugly misleading behavior:

The tx creation process was having a workaround patch to automatically
fall back to select coins from the wallet if `m_allow_other_inputs=false`
(previous default value) and no manual inputs were selected.

This could be seen in master in flows like `sendtoaddress`, `sendmany`
and even the GUI, where the `m_allow_other_inputs` value isn't customized
and the wallet still selects and adds coins to the tx internally.

furszy committed:
wallet: skip manually selected coins from 'AvailableCoins' result

No need to walk through the entire wallet's txes map just to get
coins that we could have gotten by just doing a simple map.find(out.hash).
(Which is what we are doing inside `SelectCoins` anyway)

furszy committed:
wallet: remove fetch pre-selected-inputs responsibility from SelectCoins

so if there is an error in any of the pre-set coins, we can fail right away
without computing the wallet available coins set (calling `AvailableCoins`)
which is a slow operation as it goes through the entire wallet's txes map.

----------------------

And to make the Coin Selection flow cleared, have decoupled SelectCoins in two functions:

1) AutomaticCoinSelection.
2) SelectCoins.

1) AutomaticCoinSelection:
   Receives a set of coins and selects the best subset of them to
   cover the target amount.

2) SelectCoins
   In charge of select all the user manually selected coins first ("pre-set inputs"), and
   if coin_control 'm_allow_other_inputs=true', call 'AutomaticCoinSelection' to select a
   subset of coins owned by the wallet to cover for the target - preset_inputs.total_amount
   remaining value.

furszy commented:
Took me a bit but, while was tackling the feedback, ended up re-organizing the commits changes better. The rationale for each modification should be clearer now.

What did was:

1) Eliminated the pre-set inputs second iteration inside `SelectCoins` by adding an `AddInputs(std::vector<COutput>&)` method to `SelectionResult` (so we don't need to iterate over the inputs just to insert them inside an ephemeral `OutputGroup` whom is not used for anything --> `SelectionResult::AddInput(OutputGroup&)` takes directly the `m_outputs` internal field..).

2) Removed the `FetchCoins` wrapper function and the `CoinsFund` struct:
    Moved the internals to `CreateTransactionInternal` who will first call `FetchSelectedInputs` then call `AvailableCoins`.

3)  Decoupled the SelectCoins into two functions:

    * `AutomaticCoinSelection`:
Which is basically the previously called SelectCoins without the manually selected inputs stuff.
So, it will purely receive a set of coins and be in charge of selecting the best subset of them to
cover the target amount.

    * `SelectCoins`:
In charge of selecting all the user manually selected coins first ("pre-set inputs"), and
if coin_control 'm_allow_other_inputs=true', call 'AutomaticCoinSelection' to select a
set of coins such that the target - pre_set_inputs.total_amount is met. Then merge
both results.

4) Have set `CoinControl::m_allow_other_inputs` default value to `true` to correct
a misleading behavior (convo started [here](https://github.com/bitcoin/bitcoin/pull/25685#discussion_r938977841)):
    - As several flows in the wallet such as `sendtoaddress`, `sendmany`
and even the GUI don't change the `m_allow_other_inputs` param,
and the default value for it has been `false`. The process was having 
a workaround patch to automatically fall back to select coins from the
wallet if `m_allow_other_inputs=false` and no manual inputs were selected.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -105,6 +105,61 @@ void CoinsResult::clear()
+    std::set<COutput> coins;
+    // If subtract fee from outputs is active, the 'total_amount'
+    // will be the sum of each output effective value
+    // instead of the sum of the outputs amount
```
In 4946384e2f56333274baf7fab1dfd41b046e12a1 "wallet: encapsulate pre-selected-inputs lookup into its own function"

This comment is backwards. The total amount uses the effective value normally, and real value when SFFO.

furszy replied:
ups, pushed.

achow101 commented on src/wallet/spend.h:
```diff
@@ -136,9 +135,36 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
+std::optional<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, CoinsResult& available_coins, const CAmount& nTargetValue, const CCoinControl& coin_control,
                  const CoinSelectionParams& coin_selection_params) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
 
+struct SelectedInputs
+{
+    std::set<COutput> coins;
+    // If subtract fee from outputs is active, the 'total_amount'
+    // will be the sum of each output effective value
+    // instead of the sum of the outputs amount
+    CAmount total_amount{0};
+
+    void Insert(const COutput& output, bool subtract_fee_outputs)
+    {
+        if (subtract_fee_outputs) {
+            total_amount += output.txout.nValue;
+        } else {
+            total_amount += output.GetEffectiveValue();
+        }
+        coins.insert(output);
+    }
+};
```
In bc16c074cac96e2ee9a2f3e5d2705b6168f76f80 "wallet: remove fetch pre-selected-inputs responsibility from SelectCoins"

ISTM This should just be in the previous commit.

furszy replied:
done

furszy commented:
thanks achow101, feedback tackled.

josibake commented on src/wallet/spend.cpp:
```diff
@@ -105,6 +105,47 @@ void CoinsResult::clear()
 
+// Fetch and validate the coin control selected inputs.
+// Coins could be internal (from the wallet) or external.
+util::Result<SelectedInputs> FetchSelectedInputs(const CWallet& wallet, const CCoinControl& coin_control,
```
I'd recommend calling this `GetPreSelectedInputs`. we tend to use `Get*` throughout the codebase and since this function is for pre-selected inputs only, it makes sense to make the function name specific

furszy replied:
The rationale here to use `Fetch*` instead of `Get*` was that we aren't only accessing wallet's UTXOs. We are fetching external UTXO provided by the coin control object as well.

So, with the `Fetch*` prefix aimed to make that distinction clear.
If the function would be only getting wallet internal data, then yeah, this should had been a `Get*` for sure.

josibake commented on src/wallet/spend.h:
```diff
@@ -125,6 +125,26 @@ std::optional<SelectionResult> AttemptSelection(const CWallet& wallet, const CAm
                         const CoinSelectionParams& coin_selection_params);
 
+// User manually selected inputs that must be part of the transaction
+struct SelectedInputs
```
why not call this `PreSelectedInputs` or `ManualInputs`? `SelectedInputs` doesn't really tell me who selected them and could easily be confused with a `SelectionResult`

furszy replied:
sure

josibake reviewed (commented):
Concept ACK

I really like that this simplifies `SelectCoins` to "give me coins I am allowed to spend and I will select a subset to fund this transaction." Also, +1 for the fail-fast if we are trying to use pre-selected coins and encounter an error. Thanks for taking the time to benchmark this, too. Makes a compelling case for the performance benefits.

Will do a more thorough code review tomorrow but left some initial feedback

josibake commented:
Regarding the bench, it would be nice to have a benchmark for when a user has a few pre-selected inputs and _also_ uses automatic coin selection. My hunch is that that is the more representative scenario

aureleoules reviewed (commented):
Thanks for the all documentation you've provided about this PR @furszy, it made the review easier to understand.

4cb629f3df8d3d6ea6ba0d1560b5cc3caf4ee01a:
I verified that `AvailableCoins` returned an empty `CoinsResult` if `m_allow_other_inputs` is false, so no need to loop through coins.

65d8ba52e828a3c44b962db2f1f8e230b09aab88:
In my understanding, this commit allows `AvailableCoins` to retrieve only coins that are not already pre-selected by the user, which makes the two sets of coins disjoints. Thus there is no need to remove pre-selected coins from `available_coins` anymore.

As a side note, I think it would be great to add comments to the functions below to explain that these are user pre-selected coins.
https://github.com/bitcoin/bitcoin/blob/243d7bde7859777ebe9442477da1b77419bd3f32/src/wallet/coincontrol.h#L66-L74
Same here 
https://github.com/bitcoin/bitcoin/blob/243d7bde7859777ebe9442477da1b77419bd3f32/src/wallet/coincontrol.h#L135

a53748e3323a33829134cb575830933ac1893323:
I verified `FetchSelectedInputs` is only a wrapping of the pre-selected-inputs lookup from `AvailableCoins` with a better error handler.

e94e9ce68dc2fdb06a2da17020e067baede72d1e:
I believe this commit makes the Coin Selection flow much easier to read and also avoids iterating over the wallet's tx map, which increases performance.

nits in commit description:
"we can fail right away without ~~compute~~ computing the wallet..."
"In charge of ~~select~~ selecting all the user manually selected coins..."

3e5bfcc75db1f2abf875fbcd1ed68f8166683ea4:
Verified that if there are sufficient funds with the manually selected coins, there is no need to call the automatic coin selection.

I ran the benchmark both on master and this PR and got similar results as the PR description!
I would agree with @josibake regarding having a benchmark for "when a user has a few pre-selected inputs and also uses automatic coin selection" though.

furszy commented on src/wallet/spend.cpp:
```diff
@@ -885,17 +878,29 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
+
+    // Fetch wallet available coins if "other inputs" are
+    // allowed (coins automatically selected by the wallet)
+    CoinsResult available_coins;
+    if (coin_control.m_allow_other_inputs) {
+        available_coins = AvailableCoins(wallet,
+                                         &coin_control,
+                                         coin_selection_params.m_effective_feerate,
+                                         1,            /*nMinimumAmount*/
+                                         MAX_MONEY,    /*nMaximumAmount*/
+                                         MAX_MONEY,    /*nMinimumSumAmount*/
+                                         0);           /*nMaximumCount*/
+    }
```
Yeah, it would.

But, I'm a ~0 there, as `CoinsResult` will only have a map and a `CAmount` after #25734, this will make the code a bit harder to read for a little benefit (`AvailableCoins` has too many params so you could easily get confused by one of them and don't read the `CoinsResult` default initialization at the end).

Probably with something like da44df91 the readability wouldn't be a problem anymore. And we could initialize `CoinsResults` in a single line.

furszy commented on src/wallet/spend.cpp:
```diff
@@ -529,75 +571,37 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
+
+std::optional<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, CoinsResult& available_coins, const CAmount& nTargetValue, const CCoinControl& coin_control, const CoinSelectionParams& coin_selection_params)
+{
+    CAmount value_to_select = nTargetValue;
```
yeah great, less changes if we use `value_to_select` instead. (`value_to_select` is used in every `AttemptSelection` call)

furszy commented on src/wallet/spend.cpp:
```diff
@@ -885,17 +878,29 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
-                                              MAX_MONEY,    /*nMaximumAmount*/
-                                              MAX_MONEY,    /*nMinimumSumAmount*/
-                                              0);           /*nMaximumCount*/
+    // Fetch manually select coins
```
pushed

furszy commented:
Thanks for the review guys :smile:.

> Regarding the bench, it would be nice to have a benchmark for when a user has a few pre-selected inputs and also uses automatic coin selection. My hunch is that that is the more representative scenario

You know @josibake, I didn't add it here mainly because this PR improves a particular scenario; when the user pre-selected inputs and disallowed the wallet automatic coin selection.
This PR doesn't improve, at least not directly, the preset inputs + automatic coin selection scenario.

But sure, will add another bench case for it :). Was planning to push it on a follow-up work to measure other changes but can sneak it here as well.

> As a side note, I think it would be great to add comments to the functions below to explain that these are user pre-selected coins.

Agree @aureleoules, shouldn't be called `setSelected` and have some comments there. But would say to tackle it in a separate PR so the entire `CCoinControl` class receives some love.

furszy commented:
As requested, added a basic benchmark for the preset inputs + automatic coin selection scenario in 631f766.
In future PRs, we could continue expanding the bechmarks, adding cases for different parts of the transaction creation process.

furszy commented:
rebased on master, conflict solves. Ready to go

furszy commented:
rebased, conflicts solved.

furszy committed:
wallet: SelectCoins, return early if target is covered by preset-inputs

furszy commented:
rebased, conflicts solved. Wasn't so simple this time.

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/25685/commits/08012f03bf2cc9ec4f8b21d5643f3423acb1b5d8

furszy commented:
Friendly ping to previous reviewers @aureleoules @josibake @achow101.
Would be nice to move forward with this one, it will unlock #25806 which will unlock #24845.

achow101 commented on src/wallet/coinselection.cpp:
```diff
@@ -444,6 +444,12 @@ void SelectionResult::AddInput(const OutputGroup& group)
     m_use_effective = !group.m_subtract_fee_outputs;
 }
 
+void SelectionResult::AddInputs(const std::set<COutput>& inputs, bool m_subtract_fee_outputs)
```
In e6b9e23699a5dc8e08711004f51c877fa3912db1 "wallet: encapsulate pre-selected-inputs lookup into its own function"

nit: only member variables should have the `m_` prefix.

furszy replied:
yeah, removing it..

achow101 commented on src/bench/wallet_create_tx.cpp:
```diff
@@ -115,6 +132,11 @@ static void WalletCreateTx(benchmark::Bench& bench, const OutputType output_type
+static void WalletCreateTxUsePresetInputsAndCoinSelection(benchmark::Bench& bench) { WalletCreateTx(bench, OutputType::BECH32, /*allow_other_inputs=*/true,
+                                                                                                    {{/*num_of_internal_inputs=*/4}}); }
 
-BENCHMARK(WalletCreateTxUseOnlyPresetInputs);
```
In 08012f03bf2cc9ec4f8b21d5643f3423acb1b5d8 "bench: add benchmark for tx creation using pre-set inputs and coin selection"

nit: This should be in the previous commit.

furszy commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1137,7 +1137,7 @@ RPCHelpMan send()
                 Cat<std::vector<RPCArg>>(
                 {
-                    {"add_inputs", RPCArg::Type::BOOL, RPCArg::Default{false}, "If inputs are specified, automatically include more if they are not enough."},
+                    {"add_inputs", RPCArg::Type::BOOL, RPCArg::Default{true}, "If inputs are specified, automatically include more if they are not enough."},
```
This is more a bugfix than a behavior change actually. Let me expand:

Before, the default value was not `false` all the time, it was a hidden `true`. We were having a bug/workaround to always select coins automatically from the wallet even when the user explicitly set `add_inputs=false` (which means "don't select coins automatically").

This is why, even when I changed the field default value inside coin control, I didn't need to change any test on that commit. They actually were all depending on the bug to work properly! (including the GUI..).

Can try it by setting `add_inputs=false` in any of the fund* RPC commands without pre-setting any input. The process will automatically select coins from the wallet even when you are explicitly blocking it (the proper expected behavior is fail due "insufficient funds").
I'm going to add test coverage for this.. so the previous bad behavior and the fix commit can be contrasted. It truly was a misleading behavior.

Side from that:
I know what you mean, if the user pre-set inputs, the previous default value was `false`. So, to not change behavior in this scenario, by default, no automatic coin selection should be performed on the RPC commands 👍🏼 .
Will add coverage for this as well.
Thanks.

furszy replied:
Pushed #26053 to fulfill this purposes, make the bug visible and to try to fix it for 24.0.
Once that one gets merged, changes in 0d6da166 will be smaller and simpler to review.

furszy commented:
Rebased on master, achow101 feedback tackled. Thanks.

Changes:
* Removed `m_` prefix from `AddInputs` function signature.
* Squashed benchmark 08012f03 inside 2d9ea15.
* Added test case for when the user explicitly disallows the automatic coin selection and does not provide inputs (on master, this test fails because the wallet invalidly fall back to select coins automatically even when the user provided `add_inputs=false`).

aureleoules reviewed (commented):
ACK 2d9ea15fe9773c3efe349579413b4a4e56c30454.
Reviewed with range-diff since my last review. Also verified all commits compile.

aureleoules commented:
Here are my bench results (with `sudo pyperf system tune`).

## PR
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|        3,165,243.33 |              315.93 |    1.6% |   18,936,699.67 |    7,255,808.00 |  2.610 |     507,961.67 |    2.3% |      0.10 | `WalletCreateTxUseOnlyPresetInputs`
|    2,503,088,950.33 |                0.40 |    2.9% |12,628,647,472.00 |5,467,638,592.00 |  2.310 |2,036,987,578.67 |    0.3% |     84.82 | `WalletCreateTxUsePresetInputsAndCoinSelection`


## Master
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       66,423,839.67 |               15.05 |    0.4% |  102,368,093.00 |  152,193,632.00 |  0.673 |  22,866,396.33 |    0.6% |      2.20 | `WalletCreateTxUseOnlyPresetInputs`
|    2,546,070,421.67 |                0.39 |    5.5% |12,710,666,931.00 |5,631,797,760.00 |  2.257 |2,052,977,444.33 |    0.3% |     86.15 | :wavy_dash: `WalletCreateTxUsePresetInputsAndCoinSelection` (Unstable with ~3.0 iters. Increase `minEpochIterations` to e.g. 30)

> You know josibake, I didn't add it here mainly because this PR improves a particular scenario; when the user pre-selected inputs and disallowed the wallet automatic coin selection.
This PR doesn't improve, at least not directly, the preset inputs + automatic coin selection scenario.

Indeed :+1:
`WalletCreateTxUseOnlyPresetInputs` is significantly faster now and `WalletCreateTxUsePresetInputsAndCoinSelection` has not changed.

> Unstable with ~3.0 iters. Increase `minEpochIterations` to e.g. 30

nit: maybe increase iterations? not sure how important this is.

achow101 commented on src/bench/wallet_create_tx.cpp:
```diff
@@ -0,0 +1,142 @@
+
+    // Generate chain; each coinbase will have two outputs to fill-up the wallet
+    const auto& params = Params();
+    unsigned int chain_size = 150000; // 150k blocks means 300k UTXO for the wallet (minus 200 due COINBASE_MATURITY)
```
In 2d9ea15fe9773c3efe349579413b4a4e56c30454 "bench: benchmark transaction creation process"

150000 blocks is excessive. This makes the benchmark extremely slow to run. It increases the runtime of `make check` by several orders of magnitude. A few hundred blocks should be enough to illustrate the performance difference, I don't think we need more than a thousand, let one 150,000.

furszy replied:
I initially set it low, at 2k (4k utxo), but it wasn't really illustrative. This becomes a problem when your wallet has a high tx volume. Otherwise, traverse 4k utxo is quite fast, almost imperceptible.
e.g.
### PR
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|        1,427,966.60 |              700.30 |    1.2% |      0.08 | `WalletCreateTxUseOnlyPresetInputs`

### Master
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|        1,576,250.00 |              634.42 |    0.5% |      0.09 | `WalletCreateTxUseOnlyPresetInputs`
----

Let me try with 20k instead. We should try to get a middle-ground between the high value (which I agree that it's too high) and a number that can provide meaningful information on the context that this becomes a real problem for the user.

achow101 replied:
20k is still a lot and the setup does still take quite a long time. While the real time difference may not be that noticeable, I think `ns/op` and `op/s`

furszy replied:
I'm not so convinced there. I do understand the `make check` time concern but that is limiting our testing capabilities. Some big issues aren't problems at all in a small scale.

As time isn't constant between runs and the difference is so small with 2000 blocks, stuff like this can easily go unnoticed for a really long time. Which kinda break the purpose of having a benchmark for it.

E.g. lets say that we have this benchmark in the code with the 2000 blocks param. And someone introduces a similar problematic (traverse the whole wallet map and do other stuff for no reason).
Would you care about it if the benchmark returns a time diff between 60-100 ns? or would you just say "it's ok"?.
We could be hurting badly big wallets with our changes without noticing it at all.

Saying that, will decrease the number to 10k blocks. So we get closer to each other number and can move forward with it.
I think that the 2k blocks number is too small to be helpful on this context.

----------------------

As a future thought, would be nice to be able to re-use setups in the benchmarks. So every `WalletCreateTx*` would not require to create a new chain per run.

achow101 replied:
Even with 2000 blocks I am still able to measure a difference. It's also a lot more noticeable (and everything is a lot slower) if configured with `--enable-debug`.

furszy replied:
Interesting, that is why we have been seeing different pictures of this. Without the debug mode, the difference that have been seeing locally is almost imperceptible with 2k (less than 100 ns).
That explains why the CI was timing out before, benchmarks there are running on debug mode as well.

furszy commented:
Updated per feedback, thanks aureleoules and achow101.

1) Per aureleoules review: 
   Increased the bench iterations from 3 to 5.

2) Per achow101 review: 
   Decreased the number of blocks used in the benchmarks from 150k to 20k so `make check` doesn't take too much time on low-end devices and we still continue benchmarking the tx creation process on a wallet with a reasonable amount of txes/utxo.

aureleoules reviewed (commented):
re-ACK 1c7417711d2b9353350a8b5b2e580850cd1b5d3b

aureleoules commented:
ACK 40452b8be29c85fbd2e96f93fa1b9ad37e6d44d8

## PR
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|        3,252,252.20 |              307.48 |    0.7% |   19,062,206.80 |    7,444,723.20 |  2.560 |     504,438.20 |    2.5% |      0.18 | `WalletCreateTxUseOnlyPresetInputs`

## Master
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|        7,162,568.40 |              139.61 |    1.1% |   24,654,767.60 |   16,410,489.60 |  1.502 |   1,935,246.40 |    1.3% |      0.39 | `WalletCreateTxUseOnlyPresetInputs`

furszy commented:
rebased, conflicts solved.

furszy commented:
Updated per feedback.
Only changed the benchmark chain size, from 10k to 5k. Tiny [diff](https://github.com/bitcoin/bitcoin/compare/f22a671f095780a1e0e296aab2c4d5464d0f683f..044953de5905d722fe07d0fb529125574bdb4902).

aureleoules reviewed (approved):
re-ACK 044953de5905d722fe07d0fb529125574bdb4902 - only reduced benchmark chain size since last my ACK

furszy committed:
wallet: simplify preset inputs selection target check

we are already computing the preset inputs total amount inside `PreSelectedInputs::Insert`,
which internally decides whether to use the effective value or the raw output value based on
the 'subtract_fee_outputs' flag.

furszy commented:
Rebased after #26203 merge.

Plus, to be clear over https://github.com/bitcoin/bitcoin/pull/26203#pullrequestreview-1128801074, have decoupled what previously was in aa5fdbad in two commits: 71f5817 and 8326d17.

Ready to go.

theStack commented:
Concept ACK

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -575,11 +575,12 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& a
+        if (selection_target > 0) return std::nullopt; // fails only if 'allow_other_inputs=false'
         SelectionResult result(nTargetValue, SelectionAlgorithm::MANUAL);
         result.AddInputs(pre_set_inputs.coins, coin_selection_params.m_subtract_fee_outputs);
         result.ComputeAndSetWaste(coin_selection_params.min_viable_change, coin_selection_params.m_cost_of_change, coin_selection_params.m_change_fee);
```
in d54606dad097ac050dfdd1c820e3d7172b036146
IMO this boolean logic:
```cpp
    // If automatic coin selection was disabled or if we already covered the
    // target with the preset inputs: try to return the result right away
    if (!coin_control.m_allow_other_inputs && selection_target > 0) return std::nullopt;

    if (selection_target <= 0) {
        SelectionResult result(nTargetValue, SelectionAlgorithm::MANUAL);
        result.AddInputs(pre_set_inputs.coins, coin_selection_params.m_subtract_fee_outputs);
        result.ComputeAndSetWaste(coin_selection_params.min_viable_change, coin_selection_params.m_cost_of_change, coin_selection_params.m_change_fee);
        return result;
    }
```
is easier to understand than this one:
```cpp
    // If automatic coin selection was disabled or if we already covered the
    // target with the preset inputs: try to return the result right away
    if (!coin_control.m_allow_other_inputs || selection_target <= 0) {
        // 'selection_target' is computed on `PreSelectedInputs::Insert` which decides whether to use the effective value
        // or the raw output value based on the 'subtract_fee_outputs' flag.
        if (selection_target > 0) return std::nullopt; // fails only if 'allow_other_inputs=false'
        SelectionResult result(nTargetValue, SelectionAlgorithm::MANUAL);
        result.AddInputs(pre_set_inputs.coins, coin_selection_params.m_subtract_fee_outputs);
        result.ComputeAndSetWaste(coin_selection_params.min_viable_change, coin_selection_params.m_cost_of_change, coin_selection_params.m_change_fee);
        return result;
    }
```

furszy replied:
yeah, better.

aureleoules reviewed (commented):
Left a comment, otherwise LGTM.

furszy commented:
Thanks @aureleoules, feedback applied.

Simplified the `SelectCoins` early return logic. Small [diff](https://github.com/bitcoin/bitcoin/compare/176969420e94e720cd9c4ee94520db923bc3204e..1c137b577130b9669db9a7b167f44f6304f7df42).

aureleoules reviewed (approved):
re-ACK 1c137b577130b9669db9a7b167f44f6304f7df42

furszy commented:
Rebased after #26158 merge.

Only diff is inside the benchmarks, where both of them were marked as `low` priority to not run them in `make check`.

Ready for another review round.

achow101 commented:
ACK 5bea640cfb29a7d4e2edaa2a24c0457dbda41b0a

aureleoules reviewed (approved):
reACK 5bea640cfb29a7d4e2edaa2a24c0457dbda41b0a

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -575,11 +575,11 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& a
+
+    // Return if we can cover the target only with the preset inputs
+    if (selection_target <= 0) {
         SelectionResult result(nTargetValue, SelectionAlgorithm::MANUAL);
```
nit: I'd suggest to create `SelectionResult` for preselected inputs unconditionally. This way use can reuse it later when you need to merge it with automatic coin selection.

furszy replied:
I thought about it but It's not that straightforward. The `SelectionResult` target is provided on the constructor and then only modified by the `Merge` function (which adds the merged selection target to the base one).

So, it's not clear with which target the selection result should be initialized:

1) If the preset inputs cover the target amount, then it needs to be initialized to the target amount, not to the preset inputs total amount. So the selection result can be returned early.

2) If the preset inputs don't cover the target amount, then it needs to be initialized to the preset inputs total amount, because we need to cover the total amount when we merge it with the `SelectionResult` that comes from `AutomaticCoinSelection` (which contains a selection for the target amount minus the preset inputs total).

S3RK replied:
Yap, agreed.

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -594,9 +594,6 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& a
     preselected.AddInputs(pre_set_inputs.coins, coin_selection_params.m_subtract_fee_outputs);
     op_selection_result->Merge(preselected);
-    if (op_selection_result->GetAlgo() == SelectionAlgorithm::MANUAL) {
-        op_selection_result->ComputeAndSetWaste(coin_selection_params.min_viable_change, coin_selection_params.m_cost_of_change, coin_selection_params.m_change_fee);
```
nit: I'd we need to recalculating waste after adding preselected inputs because otherwise wronge waste score would be passed to USDT

furszy replied:
funny that I mentioned this here https://github.com/bitcoin/bitcoin/pull/25685#discussion_r938067217 (last sentence) almost 3 months ago.

Will re-check it.

S3RK reviewed (commented):
I'm a bit late to the party. Almost code review ACK. Need a bit more time to review. Left a couple nits. The biggest one being about whether we need a new struct about preselected inputs.

furszy commented on src/wallet/spend.h:
```diff
@@ -121,6 +121,26 @@ std::optional<SelectionResult> AttemptSelection(const CWallet& wallet, const CAm
                         const CoinSelectionParams& coin_selection_params);
 
+// User manually selected inputs that must be part of the transaction
+struct PreSelectedInputs
```
We actually had this convo at the beginning of the PR (https://github.com/bitcoin/bitcoin/pull/25685#discussion_r938067217).

I implemented it two/three months ago and the end result wasn't so satisfactory.

If my memory doesn’t fail me, the reasons were:

1. We would be mixing responsibilities if we do that. The `SelectionResult` class is particular to the coin selection process (in the future, it might even be decoupled into a separated library) and `FetchSelectedInputs` is a wallet method that we use to obtain a set of outputs (similar to `AvailableCoins` but instead of traversing the whole wallet tx map, it fetches the specific outputs directly), which could be used in different contexts and not only inside the tx creation process (places where the `SelectionResult` class wouldn't make much sense).

2. The `SelectionResult` constructor requires a target, which we don’t have up until we fetch and sum-up all the manually selected outputs. So, we would end up doing the same that we do inside `SelectCoins` but inside `FetchSelectedInputs`: (1) first fetch and store all the outputs in a vector, then at the end of the method, create a `SelectionResult` with the accurate target and return it (so there would really be no difference with the current approach unless we add another constructor and a new `AddInput` method inside `SelectionResult` that keeps track of target etc or.. pass the tx target to `FetchSelectedInputs` which I think that is ugly as the tx target has nothing to do with a fetching-only function). 
3. Plus, the new struct caches the total amount and contemplates the effective/raw amount distinction. Which, for example, made #26203 issue a non-issue here.
.

With this I’m not saying that we should marry to this struct (we could also use `CoinsResult` instead) but I don’t think that changing it to `SelectionResult` will make the code better (at least for now). It would probably just continue entangling the coin selection code inside wallet general functions that we should be able to re-use in other processes.

Plus, after the next PR in the line #25806, I think that this process will be in a much better shape to re-do this convo once more. Mainly because the coin selection process will not be so attached to the wallet internals. So, we might be able to decouple it into a different module or an external library.

furszy commented:
Updated per @S3RK feedback, thanks.

Only change is the `ComputeAndSetWaste` call after the merge of the two selection results (preset and automatic selection results). So the waste score is properly calculated for the USDT logging. Small [diff](https://github.com/bitcoin/bitcoin/compare/5bea640cfb29a7d4e2edaa2a24c0457dbda41b0a..3fcb545ab26be3e785b5e5654be0bdc77099d827)

aureleoules reviewed (approved):
reACK 3fcb545ab26be3e785b5e5654be0bdc77099d827

S3RK commented:
Code Review ACK 3fcb545ab26be3e785b5e5654be0bdc77099d827

achow101 commented:
ACK 3fcb545ab26be3e785b5e5654be0bdc77099d827
